internal/provider/resource_gitlab_project_runner_enablement.go

```diff
@@ -0,0 +1,144 @@
+package provider
+
+import (
+	"context"
+	"log"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+var _ = registerResource("gitlab_project_runner_enablement", func() *schema.Resource {
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_project_runner_enablement`" + ` resource allows to enable a runner in a project.
+
+**Upstream API**: [GitLab REST API docs](https://docs.gitlab.com/ee/api/runners.html#enable-a-runner-in-project)`,
+		CreateContext: resourceGitlabProjectRunnerEnablementCreate,
+		ReadContext:   resourceGitlabProjectRunnerEnablementRead,
+		DeleteContext: resourceGitlabProjectRunnerEnablementDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Description: "The ID or URL-encoded path of the project owned by the authenticated user.",
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+			},
+			"runner_id": {
+				Description: "The ID of a runner to enable for the project.",
+				Type:        schema.TypeInt,
+				ForceNew:    true,
+				Required:    true,
+			},
+		},
+	}
+})
+
+func resourceGitlabProjectRunnerEnablementCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	projectID := d.Get("project").(string)
+	runnerID := d.Get("runner_id").(int)
+	options := &gitlab.EnableProjectRunnerOptions{
+		RunnerID: runnerID,
+	}
+
+	log.Printf("[DEBUG] create gitlab project runner %v/%v", projectID, runnerID)
+
+	_, _, err := client.Runners.EnableProjectRunner(projectID, options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	runnerIDString := strconv.Itoa(runnerID)
+	d.SetId(buildTwoPartID(&projectID, &runnerIDString))
+
+	return resourceGitlabProjectRunnerEnablementRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectRunnerEnablementRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	projectID, runnerID, err := projectAndRunnerFromID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	options := &gitlab.ListProjectRunnersOptions{
+		ListOptions: gitlab.ListOptions{
+			PerPage: 20,
+			Page:    1,
+		},
+		Type: gitlab.String("project_type"),
+	}
+
+	log.Printf("[DEBUG] read gitlab project runner %s/%v", projectID, runnerID)
+
+	// We're enumerating all the runners for a project to make sure the runner
+	// is indeed associated with the project.
+
+	found := false
+	for options.Page != 0 && !found {
+		runners, resp, err := client.Runners.ListProjectRunners(projectID, options, gitlab.WithContext(ctx))
```

Comment:
Instead of doing this listing & pagination, did we consider just querying the runner details and check the project assignment there? See https://docs.gitlab.com/ee/api/runners.html#get-runners-details

Reply:
Should be fairly easy to test now that we have test cases, I'll have a look

Reply:
I've changed it, but I've put it in a separate commit so you can compare. It actually still needs a call to the project, because the `project` attribute can be both a name and a numeric id, so I figured it's best to let the API figure out which it is at this point - and then use the resulting ID to compare the runner details. I figure it's error prone to test if it is numeric and compare both the id and the path in the runner details.

Have a look at the commit and see what option you like best :)

Reply:
> It actually still needs a call to the project, because the project attribute can be both a name and a numeric id, so I figured it's best to let the API figure out which it is at this point - and then use the resulting ID to compare the runner details.

Good catch! 

I still prefer the additional "constant" call to the project API to get the id for sure ... but don't have hard feelings for either of the two options.

Btw. kudos for the explicit separate commit - really helps the review 🎉

Reply:
If you prefer the new version we'll use that :)